macOS/DuckDuckGo/Sync/DeviceSyncCoordinator.swift

```diff
@@ -0,0 +1,164 @@
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import DDGSync
+import Combine
+import Common
+import SystemConfiguration
+import SyncUI_macOS
+import SwiftUI
+import Navigation
+import PixelKit
+import os.log
+import BrowserServicesKit
+
+protocol SyncDeviceFlowLaunching {
+    @MainActor
+    func startDeviceSyncFlow(completion: (() -> Void)?)
+}
+
+protocol DeviceSyncCoordinationDelegate: AnyObject {
+    @MainActor
+    func didEndFlow()
+}
+
+final class DeviceSyncCoordinator {
+    var cancellable: AnyCancellable?
+
+    @MainActor
+    init?(managementDialogModel: ManagementDialogModel = .init(), dialogController: SyncDialogController? = nil) {
+        self.managementDialogModel = managementDialogModel
+        guard let dialogController = dialogController ?? Self.createDialogController(managementDialogModel: managementDialogModel) else {
+            return nil
+        }
+        self.dialogController = dialogController
+        dialogController.coordinationDelegate = self
+    }
+
+    private let managementDialogModel: ManagementDialogModel
+    private let dialogController: SyncDialogController
+    private var syncWindowController: NSWindowController?
+
+    @MainActor
+    private func presentDialog(completion: (() -> Void)? = nil) {
+        guard !(syncWindowController?.window?.isVisible ?? false) else {
+            return
+        }
+
+        guard [AppVersion.AppRunType.normal, .uiTests].contains(AppVersion.runType) else {
+            return
+        }
+
+        let syncViewController = SyncManagementDialogViewController(managementDialogModel, dialogController: dialogController, coordinator: self)
+        syncWindowController = syncViewController.wrappedInWindowController()
+
+        guard let syncWindow = syncWindowController?.window,
+              let parentWindowController = Application.appDelegate.windowControllersManager.lastKeyMainWindowController
+        else {
+            assertionFailure("Sync: Failed to present SyncManagementDialogViewController")
+            return
+        }
+        parentWindowController.window?.beginSheet(syncWindow) { _ in
+            completion?()
+        }
+    }
+
+    @MainActor
+    private static func createDialogController(managementDialogModel: ManagementDialogModel) -> SyncDialogController? {
+        guard let syncService = NSApp.delegateTyped.syncService, let errorHandler = NSApp.delegateTyped.syncDataProviders?.syncErrorHandler else {
+            assertionFailure("Sync: Core dependencies not available")
```

Comment:
Just a reminder that if you think this may be happening in production, we could be sending a pixel here.

Reply:
Same as above: This isn't happening any more after a fix in the following PR. Instead, the only places we're handling the nil case are ViewModels which will just hide the Sync button in that case. Will probably never happen in production but it's a slightly more elegant fallback in the slim chance that sync dependencies were not ready yet.